### ISSUE TITLE
Parser throws undefined values

### ISSUE THREAD

**Issue by iwan-aucamp-cs** [bcherny/json-schema-to-typescript#182]
`parser.ts` does the following for some errors:

```
iwana@iwana-nb.concurrent.systems:~/xprojects/github.com/bcherny/json-schema-to-typescript
$ grep 'error' src/parser.ts
import { error, generateName, log } from './utils'
      throw error('Refs should have been resolved by the resolver!', schema)
  throw error('Supertype must have standalone name!', namedInterface)
```

`error` is a function which does not return anything - but just logs:

```javascript
export function error(...messages: any[]) {
  console.error(whiteBright.bgRedBright('error'), ...messages)
}
```

As a consequence, when an error does occur the output is rather messy:

```
$ json2ts --unreachableDefinitions=true ~/.local/tmp/broken.json
error Refs should have been resolved by the resolver! { definitions: 
   { Something: 
      { id: 'Something',
        title: 'The thing that is some',
        type: 'object',
        required: [Array],
        additionalProperties: false,
        properties: [Object],
        description: '' } },
  description: '',
  required: [],
  additionalProperties: true,
  id: 'Broken',
  title: 'The thing that is some',
  type: 'object',
  properties: 
   { foo: { type: 'string', title: 'the value of bar' },
     bar: { type: 'string', title: 'the value of foo' } },
  '$ref': '#' }
(node:15375) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'message' of undefined
    at /srv/filesystems/upheigie/projects/github.com/bcherny/json-schema-to-typescript/dist/src/cli.js:38:36
    at Generator.throw (<anonymous>)
    at rejected (/srv/filesystems/upheigie/projects/github.com/bcherny/json-schema-to-typescript/dist/src/cli.js:6:65)
    at <anonymous>
(node:15375) UnhandledPromiseRejectionWarning: Unhandled promise
[broken.txt](https://github.com/bcherny/json-schema-to-typescript/files/2328763/broken.txt)
 rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:15375) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

I'm not sure exactly what the intent is here - but it seems like `throw error` should be `throw Error` instead:

```diff
$ git diff src/cli.ts src/parser.ts
diff --git a/src/cli.ts b/src/cli.ts
index 1a50166..09270b1 100644
--- a/src/cli.ts
+++ b/src/cli.ts
@@ -30,7 +30,12 @@ async function main(argv: minimist.ParsedArgs) {
     const ts = await compile(schema, argIn, argv as Partial<Options>)
     await writeOutput(ts, argOut)
   } catch (e) {
-    process.stderr.write(e.message)
+    if ( process.env.VERBOSE ) {
+      console.error(e)
+    } else {
+      process.stderr.write(e.message)
+      process.stderr.write('\n')
+    }
     process.exit(1)
   }
 
diff --git a/src/parser.ts b/src/parser.ts
index feb89a0..3daced2 100644
--- a/src/parser.ts
+++ b/src/parser.ts
@@ -5,7 +5,8 @@ import { Options } from './'
 import { typeOfSchema } from './typeOfSchema'
 import { AST, hasStandaloneName, T_ANY, T_ANY_ADDITIONAL_PROPERTIES, TInterface, TInterfaceParam, TNamedInterface } from './types/AST'
 import { JSONSchema, JSONSchemaWithDefinitions, SchemaSchema } from './types/JSONSchema'
-import { error, generateName, log } from './utils'
+import { generateName, log } from './utils'
+import stringify = require('json-stringify-safe')
 
 export type Processed = Map<JSONSchema | JSONSchema4Type, AST>
 
@@ -142,7 +143,7 @@ function parseNonLiteral(
         type: 'UNION'
       })
     case 'REFERENCE':
-      throw error('Refs should have been resolved by the resolver!', schema)
+      throw Error(`Refs should have been resolved by the resolver! ${stringify(schema, undefined, 2)}`)
     case 'STRING':
       return set({
         comment: schema.description,
@@ -261,7 +262,7 @@ function newNamedInterface(
     return namedInterface
   }
   // TODO: Generate name if it doesn't have one
-  throw error('Supertype must have standalone name!', namedInterface)
+  throw Error(`Supertype must have standalone name! ${stringify(namedInterface,undefined,2)}`)
 }
 
 /**
```

This results in the following output

```json
Refs should have been resolved by the resolver! {
  "definitions": {
    "Something": {
      "id": "Something",
      "title": "The thing that is some",
      "type": "object",
      "required": [
        "foo",
        "bar"
      ],
      "additionalProperties": false,
      "properties": {
        "foo": {
          "type": "string",
          "title": "the value of bar"
        },
        "bar": {
          "type": "string",
          "title": "the value of foo"
        }
      },
      "description": ""
    }
  },
  "description": "",
  "required": [],
  "additionalProperties": true,
  "id": "Broken",
  "title": "The thing that is some",
  "type": "object",
  "properties": {
    "foo": {
      "type": "string",
      "title": "the value of bar"
    },
    "bar": {
      "type": "string",
      "title": "the value of foo"
    }
  },
  "$ref": "#"
}
```

**Comment by bcherny** [comment]
Hey @iwan-aucamp-cs! Can you explain a little more what exactly the difference in output is before and after your proposed change? A diff (like you did for the code) would be nice.

`throw error()` is a little strange, but it:
- Lets us have nicely formatted output (debatable, but it looked nice in my terminal when I wrote it a while ago :))
- Terminates the program
- Produces a stack trace

**Comment by iwan-aucamp-cs** [comment]
@bcherny I have created a pull request to fix this (#188) - it's slightly different from diff I pasted here.

Throwing a normal exception (as is done in the patch) should cover all three points which you raise - however the way it currently works the stack trace is not actually accurate.

The differences between current output and output with pull request is:

 - Current output:

```
error Refs should have been resolved by the resolver! { definitions: 
   { Something: 
      { id: 'Something',
        title: 'The thing that is some',
        type: 'object',
        required: [Array],
        additionalProperties: false,
        properties: [Object],
        description: '' } },
  description: '',
  required: [],
  additionalProperties: true,
  id: 'Broken',
  title: 'The thing that is some',
  type: 'object',
  properties: 
   { foo: { type: 'string', title: 'the value of bar' },
     bar: { type: 'string', title: 'the value of foo' } },
  '$ref': '#' }
(node:11454) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'message' of undefined
    at /srv/filesystems/upheigie/projects/github.com/bcherny/json-schema-to-typescript/dist/src/cli.js:79:46
    at step (/srv/filesystems/upheigie/projects/github.com/bcherny/json-schema-to-typescript/dist/src/cli.js:33:23)
    at Object.throw (/srv/filesystems/upheigie/projects/github.com/bcherny/json-schema-to-typescript/dist/src/cli.js:14:53)
    at rejected (/srv/filesystems/upheigie/projects/github.com/bcherny/json-schema-to-typescript/dist/src/cli.js:6:65)
    at <anonymous>
(node:11454) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:11454) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

 - Output with patch:
```
$ node dist/src/cli.js --unreachableDefinitions=true ~/.local/tmp/broken.json 2>&1 | tee /var/tmp/out.after
error Error: Refs should have been resolved by the resolver! { definitions: 
   { Something: 
      { id: 'Something',
        title: 'The thing that is some',
        type: 'object',
        required: [Array],
        additionalProperties: false,
        properties: [Object],
        description: '' } },
  description: '',
  required: [],
  additionalProperties: true,
  id: 'Broken',
  title: 'The thing that is some',
  type: 'object',
  properties: 
   { foo: { type: 'string', title: 'the value of bar' },
     bar: { type: 'string', title: 'the value of foo' } },
  '$ref': '#' }
    at parseNonLiteral (/home/iwana/projects/github.com/concsys/json-schema-to-typescript/dist/src/parser.js:126:19)
    at parse (/home/iwana/projects/github.com/concsys/json-schema-to-typescript/dist/src/parser.js:38:11)
    at Object.<anonymous> (/home/iwana/projects/github.com/concsys/json-schema-to-typescript/dist/src/index.js:118:104)
    at step (/home/iwana/projects/github.com/concsys/json-schema-to-typescript/dist/src/index.js:56:23)
    at Object.next (/home/iwana/projects/github.com/concsys/json-schema-to-typescript/dist/src/index.js:37:53)
    at fulfilled (/home/iwana/projects/github.com/concsys/json-schema-to-typescript/dist/src/index.js:28:58)
    at <anonymous>
```

**Comment by bcherny** [comment]
359df5b6c5dbee8f8e74d71aa557e3b1acda8d0a
